### PR TITLE
feat: inline media display

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- add media proxy config, service worker registration/build, and proxy URL helpers
- render inline media + attachments with download fallbacks
- update markdown sanitization and nginx /sw.js handling
- add proxy URL unit tests

## Testing
- npm run lint
- npm run test
- npm run build

Relates to #56

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- npm run lint (no errors)
- npm run test (passed: 6, failed: 0, skipped: 0)
- npm run build (success; warning: chunk size > 500 kB)

File: src/sw.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  type: 'SET_TOKEN';
+  token: string;
+  mediaProxyOrigin: string;
+};
+
+type ClearTokenMessage = {
+  type: 'CLEAR_TOKEN';
+};
+
+type MediaProxyMessage = SetTokenMessage | ClearTokenMessage;
```
Comment (noa-lucent) [inline src/sw.ts]: [major] **Duck typing / runtime shape checks in internal code.** `isRecord`, `parseSetTokenMessage`, and `isClearTokenMessage` all perform manual runtime shape inspection (`typeof value === 'object'`, property checks) on `event.data`.

This is acceptable here because Service Worker `message` events carry `unknown` data — **this is boundary code** receiving untrusted input from `postMessage`. However, the file mixes boundary parsing with internal logic (`handleMessage`, `shouldProxyRequest`, `buildAuthorizedRequest`) in the same flat module.

Separate the boundary parsing (message validation) from the internal handlers. For example, the `message` event listener should parse + validate, then call typed internal functions. Right now the structure is fine functionally, but consider grouping the boundary functions together with a clear comment/section to signal intent, or extracting them to a small parser module. This prevents future drift where someone adds more shape checks deeper in the logic.

This is minor enough to not block, but flagging since the pattern should stay isolated to the event handler boundary.

File: src/lib/media/proxy-url.test.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  }));
+  return import('./proxy-url');
+};
+
+afterEach(() => {
+  vi.resetModules();
+  vi.unmock('@/config');
+});
+
+describe('proxy-url', () => {
```
Comment (noa-lucent) [inline src/lib/media/proxy-url.test.ts]: [minor] **Test coverage gaps — edge cases and `parseAgynFileId` not fully exercised.** The tests cover the happy paths well, but consider adding:

1. **Empty/whitespace input**: `buildProxyUrl('')`, `buildProxyUrl('  ')` — ensures `normalizeInput` works as expected.
2. **Malformed agyn URLs**: `parseAgynFileId('agyn://notfile/123')`, `parseAgynFileId('agyn://file/')` (empty path) — ensures host and path validation.
3. **`buildProxyUrl` with `size: 0` or `size: -1`** — verifies `normalizeSize` rejects non-positive values.
4. **`buildDownloadUrl` with configured proxy** — currently only tested via agyn URLs; test with a plain `https://` URL too.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Good work overall — the architecture is clean, the SW approach is solid, and the component decomposition is thoughtful. A few things to address before merging:

**Must fix:**
- `getFileMetadata`: Remove the `response.file ?? (response as unknown as FileMetadata)` double-cast fallback. This is a speculative recovery path that defeats type safety and silently masks API contract violations.

**Should fix:**
- `proxy-url.ts`: `resolveMediaProxyBase` re-validates invariants already guaranteed by the config boundary.
- `files.ts` types: Three overlapping shapes for file metadata (`FileRecord`, `FileMetadata`, `FileMetadataWire`) — consolidate to two (canonical + wire).
- `MediaImage`: Skeleton and `` render simultaneously during loading, causing potential layout overlap.
- Test coverage for `proxy-url.test.ts` edge cases (empty input, malformed agyn URLs, non-positive sizes).

**Note:** The issue spec (#56) mentions E2E (Playwright) tests for media rendering. These aren't included here, which is fine if planned as a follow-up — just confirming that's intentional since the PR says "Relates to" rather than "Closes."

**Review by noa-lucent — DISMISSED** [review]
All 9 prior review comments are addressed. Verified each change against the original feedback:

- **`files.ts`**: Double-cast fallback removed. Clean `response.file` check with loud failure. No `as` casts — types flow through `FileMetadataWire` naturally.
- **`types/files.ts`**: Consolidated from 3 shapes to 2 (`FileRecord` canonical + `FileMetadataWire` wire). `FileMetadata` removed.
- **`proxy-url.ts`**: `resolveMediaProxyBase` no longer re-validates config invariants. `normalizeSize` uses `== null` instead of `typeof`.
- **`MediaImage.tsx`**: Clean three-way switch (`loading` / `error` / `loaded`). Off-screen `Image()` preload with proper cleanup. No more dual-render overlap.
- **`MessageAttachments.tsx`**: Defensive `.filter(Boolean)` removed; trusts typed `string[]`.
- **`sw.ts`**: Boundary/internal sections clearly separated with comments. Unified `parseMediaProxyMessage` entry point. Synthetic 401 response replaces silent unauthenticated forwarding.
- **`proxy-url.test.ts`**: 3 new edge case tests (empty/whitespace, invalid sizes, malformed agyn URLs) + `buildDownloadUrl` with https URL. Coverage is solid.

LGTM — nice cleanup.

**Review by noa-lucent — DISMISSED** [review]
Reviewed all 4 new changes. No issues.

- **`file-upload.spec.ts`**: Assertions correctly updated to match the new `MessageAttachments` rendering (filename + content type via `MediaFallback` card) instead of the removed "N attachments" text.
- **`sw-registration.ts`**: Stale SW cleanup is clean — best-effort unregister with an appropriately commented catch. `SW_SCRIPT_URL` constant is a good extraction.
- **`playwright.config.ts`**: `--ignore-certificate-errors` is standard for self-signed cert E2E environments, consistent with the existing `ignoreHTTPSErrors: true`.
- **`organization-switching.spec.ts`**: Timeout bump is reasonable and consistent with the rest of the suite.

All prior feedback remains resolved. LGTM.

**Review by noa-lucent — DISMISSED** [review]
Reviewed all 8 changed files across the 3 new commits. No issues.

**`urlTransform` fix (`MarkdownContent.tsx`):**
Clean and correct. `defaultUrlTransform` (react-markdown v9) strips non-standard protocols by default — the custom `urlTransform` short-circuits for `agyn:` and delegates the rest. Properly extracted constant.

**Unit test (`MarkdownContent.test.tsx`):**
Good coverage of the fix — verifies `agyn://file/some-uuid` survives the full markdown pipeline and reaches `MediaImage` with correct `src`.

**`data-testid` additions:**
Consistent naming convention across all media components. Both render branches (success and fallback/error) carry the wrapper test ID, which is important for reliable E2E selectors.

**E2E tests (`inline-media.spec.ts`):**
7 tests with solid helper extraction. Properly adapts assertions based on whether `MEDIA_PROXY_URL` is configured — handles proxy/no-proxy, loaded/error/unavailable states correctly. Covers: external URLs, `agyn://` protocol, multiple images, video, audio, mixed media, and markdown + media coexistence.

All prior feedback remains resolved. LGTM.

**Comment by casey-brooks** [comment]
## Test & Lint Summary (update)
- npm run lint (no errors)
- npm run test (passed: 10, failed: 0, skipped: 0)
- npm run build (success; warning: chunk size > 500 kB)

**Comment by casey-brooks** [comment]
## Test & Lint Summary (update)
- npm run lint (no errors)
- npm run test (passed: 15, failed: 0, skipped: 0)
- npm run build (success; warning: chunk size > 500 kB)

**Comment by casey-brooks** [comment]
## Test & Lint Summary (update)
- npm run lint (no errors)
- npm run test (passed: 15, failed: 0, skipped: 0)
- npm run build (success; warning: chunk size > 500 kB)

File: src/config.ts [hunk 3]
```diff
@@ -74,6 +101,11 @@ const rawApiBase = requireConfig('API_BASE_URL', readConfigValue('API_BASE_URL',
 const apiBaseUrl = deriveBase(rawApiBase, { stripApi: true });
 const socketBaseUrl = deriveBase(rawApiBase, { stripApi: true });
 
+const rawMediaProxyUrl = readConfigValue('MEDIA_PROXY_URL', 'VITE_MEDIA_PROXY_URL');
+const mediaProxyUrl = rawMediaProxyUrl
+  ? deriveBase(rawMediaProxyUrl, { stripApi: false })
+  : deriveMediaProxyUrl();
```
Comment (noa-lucent) [inline src/config.ts]: [major] **Implicit fallback silently activates security-sensitive feature.** Previously, when `MEDIA_PROXY_URL` was not set, `mediaProxyUrl` was `null` and the media proxy feature was fully disabled — no Service Worker registered, no auth token injection. The issue spec (#56) explicitly calls this out: "Media proxy URL is nullable — components must gracefully degrade when not configured."

Now, `deriveMediaProxyUrl()` silently activates the feature for any subdomain deployment by guessing `media.{domain}`. This means:

1. **The Service Worker registers and injects `Authorization: Bearer` tokens** into requests to the derived origin — even if no media proxy server exists there. If `media.{domain}` resolves to something unexpected (a different team's service, a CDN, a parking page), auth tokens leak.
2. **There is no way to disable the media proxy feature** without either deploying on localhost/IP or setting `MEDIA_PROXY_URL` to... what? There's no sentinel value to opt out.
3. **This contradicts the spec**, which treats the proxy as opt-in via config. Convention-based derivation is a different contract.

If auto-derivation is genuinely desired, it needs to be an explicit opt-in (e.g., `MEDIA_PROXY_URL=auto`) rather than a silent fallback. At minimum, document the behavior and provide a way to disable it.

Suggested fix — keep the previous null fallback and handle auto-derivation as an explicit opt-in:
```ts
const mediaProxyUrl = rawMediaProxyUrl === 'auto'
  ? deriveMediaProxyUrl()
  : rawMediaProxyUrl
    ? deriveBase(rawMediaProxyUrl, { stripApi: false })
    : null;
```

File: src/config.ts [hunk 4]
```diff
@@ -30,6 +32,31 @@ type OidcConfig = OidcConfigEnabled | OidcConfigDisabled;
...
+ * the first subdomain label with "media".
+ * Returns null when derivation is not possible (no subdomain, IP address, SSR, etc.).
+ */
+export function deriveMediaProxyUrl(): string | null {
+  if (typeof window === 'undefined') return null;
+
+  const { protocol, hostname, port } = window.location;
+
+  // Cannot derive from IP addresses
+  if (/^\d/.test(hostname) || hostname.startsWith('[')) return null;
```
Comment (noa-lucent) [inline src/config.ts]: [minor] **IPv6 bracket guard is dead code for `window.location.hostname`.** The check `hostname.startsWith('[')` never triggers because `window.location.hostname` strips brackets from IPv6 addresses (e.g., `http://[::1]:3000` → `hostname === '::1'`). The function still returns `null` correctly for IPv6 because `hostname.indexOf('.')` is `-1`, but the bracket check gives a false sense of coverage.

Either remove it (since the dot check already handles IPv6) or replace with a check that actually matches IPv6 hostnames:
```ts
if (/^\d/.test(hostname) || hostname.includes(':')) return null;
```

Also consider adding a test for IPv6: `setLocation('http://[::1]:3000')` → `null`.

File: test/e2e/inline-media.spec.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  const messageItem = page.getByTestId('chat-message').filter({ hasText: anchorText });
+  await expect(messageItem).toBeVisible({ timeout: 15000 });
+
+  return messageItem;
+}
+
+async function expectInlineImage(
+  messageItem: Locator,
+  altText: string,
+): Promise<void> {
```
Comment (noa-lucent) [inline test/e2e/inline-media.spec.ts:53]: [minor] **`expectInlineImage` filter by `hasText` won't match successfully loaded images.** The helper uses:
```ts
const image = messageItem.getByTestId('media-image').filter({ hasText: altText });
```

Playwright's `hasText` checks `element.textContent`, which does **not** include `` attribute values. This filter matches only when the image errors/is unavailable (because the error text includes the alt text, e.g., "Image failed to load: a sunset over the ocean").

Currently this works because the E2E proxy is unreachable and images always error. But if the E2E environment ever gains a working proxy, successfully loaded images would have no matching text content, and these tests would fail.

Consider filtering by a more reliable selector — e.g., match the `media-image` wrapper that contains an `` with the specific `alt` attribute:
```ts
const image = messageItem.locator('[data-testid="media-image"]').filter({
  has: page.locator(`img[alt="${altText}"], :text("${altText}")`),
});
```
Or add the alt text as a `data-alt` attribute on the wrapper div in `MediaImage`.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
The `urlTransform` fix, test IDs, E2E test simplification, and Chats.tsx cleanup all look good.

**Must fix:**
- `deriveMediaProxyUrl()` as a silent fallback changes the security posture of the app — auth tokens will be injected into requests to a convention-derived origin even when no proxy was configured. This should be opt-in (e.g., `MEDIA_PROXY_URL=auto`), not implicit.

**Should fix:**
- IPv6 bracket guard in `deriveMediaProxyUrl` is dead code — `window.location.hostname` strips brackets.
- `expectInlineImage` filter by `hasText: altText` only works when images error; it won't match successfully loaded images since `` doesn't contribute to `textContent`.

**Comment by casey-brooks** [comment]
## Test & Lint Summary (update)
- npm run lint (no errors)
- npm run test (passed: 16, failed: 0, skipped: 0)
- npm run build (success; warning: chunk size > 500 kB)

**Review by noa-lucent — APPROVED** [review]
All 3 prior comments verified against the code:

1. **Auto-derivation gated behind `'auto'` sentinel** — `MEDIA_PROXY_URL=auto` triggers `deriveMediaProxyUrl()`, explicit URL goes through `deriveBase`, absent config → `null`. Feature is now opt-in as requested.
2. **IPv6 guard fixed** — `hostname.includes(':')` replaces dead `hostname.startsWith('[')` check. IPv6 test case added.
3. **`expectInlineImage` selector fixed** — Uses `[data-alt="${altText}"]` attribute selector instead of `hasText`. `data-alt` added to both render branches of `MediaImage`. Works regardless of image load state.

All prior feedback across all review rounds remains resolved. LGTM.

**Review by vitramir — APPROVED** [review]
